website/docs/development/testing.mdx

```diff
@@ -0,0 +1,313 @@
+      existsSync: sinon.stub().returns(true),
+      readFileSync: sinon.stub().returns(newConfig),
+    };
+
+    const config = proxyquire('../src/config', {
+      fs: fsStub,
+    });
+
+    // Initialize the user config after proxyquiring to load the stubbed config
+    config.initUserConfig();
+
+    const authMethods = config.getAuthMethods();
+    expect(authMethods).to.have.lengthOf(3);
+    expect(authMethods[0].type).to.equal('local');
+    expect(authMethods[1].type).to.equal('ActiveDirectory');
+    expect(authMethods[2].type).to.equal('openidconnect');
+  });
+});
+```
+
+Core concepts to keep in mind when unit testing JS/TS modules with Chai:
+
+#### Stub internal methods to make tests predictable
+
+Functions often make use of internal libraries such as `fs` for reading files and performing operations that are dependent on the overall state of the app (or database/filesystem). Since we're only testing that the given function behaves the way we want, we **stub** these libraries.
+
+For example, here we stub the `fs` library so that "reading" the `proxy.config.json` file returns our mock config file:
+
+```js
+// Define the mock config file
+const newConfig = JSON.stringify({
+  authentication: [
+    { type: 'local', enabled: true },
+    { type: 'ActiveDirectory', enabled: true },
+    { type: 'openidconnect', enabled: true },
+  ],
+});
+
+// Create the stub for `fs.existsSync` and `fs.readFileSync`
+const fsStub = {
+  existsSync: sinon.stub().returns(true),
+  readFileSync: sinon.stub().returns(newConfig),
+};
+```
+
+This stub will make all calls to `fs.existsSync` to return `true` and all calls to `readFileSync` to return the `newConfig` mock file.
+
+Then, we use `proxyquire` to plug in the stub to the library that we're testing:
+
+```js
+const config = proxyquire('../src/config', {
+  fs: fsStub,
+});
+
+// Initialize the user config after proxyquiring to load the stubbed config
+config.initUserConfig();
+```
+
+Finally, when calling the function we're trying to test, the internal calls will automatically resolve to the values we chose.
+
+#### Setup and cleanup
+
+`before` and `beforeEach`, `after` and `afterEach` are testing constructs that allow executing code before and after each test. This allows setting up stubs before each test, making API calls, setting up the database - or otherwise cleaning up the database after test execution.
+
+This is an example from another test file (`test/addRepoTest.test.js`):
+
+```js
+before(async function () {
+  app = await service.start();
+
+  await db.deleteRepo('test-repo');
+  await db.deleteUser('u1');
+  await db.deleteUser('u2');
+  await db.createUser('u1', 'abc', 'test@test.com', 'test', true);
+  await db.createUser('u2', 'abc', 'test2@test.com', 'test', true);
+});
+
+// Tests go here
+
+after(async function () {
+  await service.httpServer.close();
+
+  await db.deleteRepo('test-repo');
+  await db.deleteUser('u1');
+  await db.deleteUser('u2');
+});
```

Comment:
Probably worth including a small sinon example and subsequent reset in after - see earlier comment about emphasizing the reset.

Reply:
Done!